base/iterators.jl

```diff
@@ -1095,4 +1094,35 @@ eltype(::Type{Stateful{T, VS}} where VS) where {T} = eltype(T)
 IteratorEltype(::Type{Stateful{T,VS}}) where {T,VS} = IteratorEltype(T)
 length(s::Stateful) = length(s.itr) - s.taken
 
+"""
+    only(x)
+
+Returns the one and only element of collection `x`, and throws an `ArgumentError` if the
+collection has zero or multiple elements.
+"""
+Base.@propagate_inbounds function only(x)
+    i = iterate(x)
+    @boundscheck if i === nothing
+        throw(ArgumentError("Collection is empty, must contain exactly 1 element")
+    end
+    (ret, state) = i
+    j = iterate(x, state)
+    @boundscheck if j !== nothing
+        throw(ArgumentError("Collection has multiple elements, must contain exactly 1 element")
+    end
+    return ret
+end
+
+# Collections of known size
+only(x::Tuple{}) = throw(ArgumentError("Tuple is empty, must contain exactly 1 element")
+only(x::Tuple{Any}) = x[1]
+only(x::Tuple) = throw(
+    ArgumentError("Tuple contains $(length(x)) elements, must contain exactly 1 element")
+)
+only(a::AbstractArray{<:Any, 0}) = @inbounds return a[]
+only(x::NamedTuple{<:Any, <:Tuple{Any}}) = first(x)
+only(x::NamedTuple) = throw(
+    ArgumentError("NamedTuple contains $(length(x)) elements, must contain exactly 1 element")
+)
```

Comment:
Not sure of style in Base, but this fits with other files as far as I can tell...

Reply:
I am tempted to say this should be a `BoundsError` or a `DimensionMismatch` ?

Reply:
- yeah, I can see the case for `BoundsError` as the error type... but on the other hand the message doesn't really feel right
```julia
julia> x = [1, 2];

julia> throw(BoundsError(x, 2))
ERROR: BoundsError: attempt to access 2-element Array{Int64,1} at index [2]
```

- Similarly, with `DimensionMismatch`, what is the "mismatch"?

- `ArgumentError` feels better to me... but i'd certainly like to here other opinions!

Reply:
fair